### PR TITLE
fix crash SemanticsObject dealloc and access the children

### PR DESCRIPTION
**Pull request by xiaoxiaowesley**
I found a crash in SemanticsObject.mm when `[SemanticsObject dealloc]` and access property `self.children`
but the `children` has called release

Here is the crash logic:

1. `[_children release]`first called in   `[SemanticsObject dealloc]` but _children not to set nil

```
- (void)dealloc {
  for (SemanticsObject* child in _children) {
    [child privateSetParent:nil];
  }
  [_children removeAllObjects];
  [_children release];
  _parent = nil;
  _container.get().semanticsObject = nil;
  [_platformViewSemanticsContainer release];
  [super dealloc];
}
```
2. After `[super dealloc]` will access  `[UIAccessibilityElementSuperCategory dealloc]` and then `[SemanticsObject accessibilityContainer]`  and access the released `_children` property, and crash finally

```
- (id)accessibilityContainer {
  if ([self hasChildren] || [self uid] == kRootNodeId) {
    if (_container == nil)
      _container.reset([[SemanticsObjectContainer alloc] initWithSemanticsObject:self
                                                                          bridge:[self bridge]]);
    return _container.get();
  }
  if ([self parent] == nil) {
    // This can happen when we have released the accessibility tree but iOS is
    // still holding onto our objects. iOS can take some time before it
    // realizes that the tree has changed.
    return nil;
  }
  return [[self parent] accessibilityContainer];
}

- (BOOL)hasChildren {
  if (_node.IsPlatformViewNode()) {
    return YES;
  }
  return [self.children count] != 0;
}

```

And the issue is here [https://github.com/flutter/flutter/issues/87247]( https://github.com/flutter/flutter/issues/87247)
[https://github.com/flutter/flutter/issues/66032](https://github.com/flutter/flutter/issues/66032)

### CODE REVIEW

**Comment by stuartmorgan** [comment]
> 1. After `[super dealloc]` will access  `[UIAccessibilityElementSuperCategory dealloc]` and then `[SemanticsObject accessibilityContainer]`

It is not safe to call method on an object for which dealloc has been called, so this is a bug; niling the ivar is papering over the issue; the correct fix would be to ensure that this call doesn't happen in the first place.

**Comment by jmagman** [comment]
> It is not safe to call method on an object for which dealloc has been called, so this is a bug; niling the ivar is papering over the issue; the correct fix would be to ensure that this call doesn't happen in the first place.

`UIAccessibilityElementSuperCategory` is in UIKit so the bug is theirs.  And unfortunately we don't know and can't rely on the stability of their `dealloc` implementation.

I wonder if this reproduces if the `children` getter is implemented instead of synthesized...

**Comment by jmagman** [comment]
> I wonder if this reproduces if the `children` getter is implemented instead of synthesized...

@xiaoxiaowesley Can you try this? Were you able to reproduce?

**Comment by xiaoxiaowesley** [comment]
> > I wonder if this reproduces if the `children` getter is implemented instead of synthesized...
> 
> @xiaoxiaowesley Can you try this? Were you able to reproduce?

Sorry I can't reproduce.But I  received many crash cases from my user.

**Comment by xiaoxiaowesley** [comment]
> > I wonder if this reproduces if the `children` getter is implemented instead of synthesized...
> 
> @xiaoxiaowesley Can you try this? Were you able to reproduce?

@jmagman I added test but the "needs tests" label still there. Does the code and test look good to you?

**Review by jmagman — CHANGES_REQUESTED** [review]
I don't think this is a complete workaround for the UIKit dealloc bug.  You've added a guard at the spot where it's crashing, but not in the other places `self.children` is called.  And we don't know and can't rely on their `dealloc` implementation.

It's unusual, but what if we waited to dealloc `_children` and `_platformViewSemanticsContainer` until after `[super dealloc]`?

```objc
- (void)dealloc {
  for (SemanticsObject* child in _children) {
    [child privateSetParent:nil];
  }
  [_children removeAllObjects];
  _parent = nil;
  _container.get().semanticsObject = nil;
  [super dealloc];

  // Due to a bug in -[UIAccessibilityElementSuperCategory dealloc]
  // that calls self.children, delay releasing ivars until after `[super dealloc]`.
  [_children release];
  [_platformViewSemanticsContainer release];
}
```
@stuartmorgan can you foresee any problems with this?

**Comment by stuartmorgan** [comment]
I looked at this a bit more. `children` isn't part of `UIAccessibilityElement`, so I was confused about how `[super dealloc]` could be calling it. According to the stack, it's not; we are, in `accessibilityContainer`. So if the bug is that iOS 9 is calling `accessibilityContainer` in our dealloc flow, we should target that specifically. E.g., by doing something very explicit by setting an `inDealloc` bool to true before calling `[super dealloc]`, and short-circuiting `accessibilityContainer` immediately if that bool is true (with a clear comment explaining why, pointing to the issue).

**Comment by xiaoxiaowesley** [comment]
> I looked at this a bit more. `children` isn't part of `UIAccessibilityElement`, so I was confused about how `[super dealloc]` could be calling it. According to the stack, it's not; we are, in `accessibilityContainer`. So if the bug is that iOS 9 is calling `accessibilityContainer` in our dealloc flow, we should target that specifically. E.g., by doing something very explicit like setting an `inDealloc` bool to true before calling `[super dealloc]`, and short-circuiting `accessibilityContainer` immediately if that bool is true (with a clear comment explaining why, pointing to the issue).

In apple's documentation UIAccessibilityElement.h said:  

> UIAccessibilityElement 
>  Instances of this class can be used as "fake" accessibility elements.

Since SemanticsObject which is the subclass of UIAccessibilityElement  maintain its own data structure like children and container in its own life cycle. So it must be release all the properties before call `[super dealloc]`. Though in iOS 9 will call `accessibilityContainer ` in  `[super dealloc]`, we should make sure all the properties(children or container) is nil before `[super dealloc]` called.

**Comment by stuartmorgan** [comment]
> we should make sure all the properties(children or container) is nil before `[super dealloc]` called.

Why? Do you have evidence that anything other than `accessibilityContainer` is improperly called during `[super dealloc]`?

**Comment by xiaoxiaowesley** [comment]
> > we should make sure all the properties(children or container) is nil before `[super dealloc]` called.
> 
> Why? Do you have evidence that anything other than `accessibilityContainer` is improperly called during `[super dealloc]`?

Of course I not sure if other override function been called during `[super dealloc]` since iOS framework is black box for us.
That is the point we should provent from access the released properties like `children` or `contianer`.

Setting all properies to nil is a way and using an `inDealloc ` bool to do this thing a way eather.

@stuartmorgan I add `_inDealloc` bool to prevent from accessing the `children` in `accessibilityContainer`.
Thanks for reviewing.

File: shell/platform/darwin/ios/framework/Source/SemanticsObjectTest.mm [hunk 1]
```diff
@@ -644,6 +644,18 @@ - (void)testFlutterSwitchSemanticsObjectMatchesUISwitch {
...
+- (void)testSemanticsObjectDeallocated {
+  fml::WeakPtrFactory<flutter::AccessibilityBridgeIos> factory(
+      new flutter::MockAccessibilityBridge());
+  fml::WeakPtr<flutter::AccessibilityBridgeIos> bridge = factory.GetWeakPtr();
+  SemanticsObject* parent = [[SemanticsObject alloc] initWithBridge:bridge uid:0];
+  SemanticsObject* child = [[SemanticsObject alloc] initWithBridge:bridge uid:1];
+  parent.children = @[ child ];
+  __weak SemanticsObject* weakObject = parent;
+  parent = nil;
+  XCTAssertNil(weakObject);
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/SemanticsObjectTest.mm:656]: The probably could use a comment that it's making sure dealloc doesn't crash, it looks like it's not testing anything other than that `__weak` is working.

**Review by jmagman — APPROVED** [review]
> > we should make sure all the properties(children or container) is nil before `[super dealloc]` called.
> 
> Why? Do you have evidence that anything other than `accessibilityContainer` is improperly called during `[super dealloc]`?

I spent all day yesterday yak shaving trying to get the engine tests to run on iOS 9 to see what else`-dealloc` could be calling.  My macOS is too new to run an iOS 9 simulator, and I actually have a 9.3.6 iPhone 4s but I can't run the engine unit tests on a physical device. https://github.com/flutter/flutter/issues/87699

This is as good a hack as any.  We can see if any new crashes surface.  LGTM

File: shell/platform/darwin/ios/framework/Source/SemanticsObject.mm [hunk 2]
```diff
@@ -399,6 +400,7 @@ - (instancetype)initWithBridge:(fml::WeakPtr<flutter::AccessibilityBridgeIos>)br
     _bridge = bridge;
     _uid = uid;
     _children = [[NSMutableArray alloc] init];
+    _inDealloc = NO;
```
Comment (stuartmorgan) [inline shell/platform/darwin/ios/framework/Source/SemanticsObject.mm:403]: Don't explicitly initialize things to their defaults values, per ObjC style guide. This line can just be removed.

**Review by stuartmorgan — APPROVED** [review]

File: shell/platform/darwin/ios/framework/Source/SemanticsObject.mm (hunk 2 above)
Comment (jmagman) [reply shell/platform/darwin/ios/framework/Source/SemanticsObject.mm:403]:
```suggestion
```

File: shell/platform/darwin/ios/framework/Source/SemanticsObjectTest.mm [hunk 3]
```diff
@@ -644,6 +644,22 @@ - (void)testFlutterSwitchSemanticsObjectMatchesUISwitch {
...
+  fml::WeakPtrFactory<flutter::AccessibilityBridgeIos> factory(
+      new flutter::MockAccessibilityBridge());
+  fml::WeakPtr<flutter::AccessibilityBridgeIos> bridge = factory.GetWeakPtr();
+  SemanticsObject* parent = [[SemanticsObject alloc] initWithBridge:bridge uid:0];
+  SemanticsObject* child = [[SemanticsObject alloc] initWithBridge:bridge uid:1];
+  parent.children = @[ child ];
+  // In some iOS 9 physical devices it will crash when access released properties during
+  // `[super dealloc]` and this test will no pass
+  // https://github.com/flutter/flutter/issues/66032
+  // https://github.com/flutter/engine/pull/27786
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/SemanticsObjectTest.mm:657]:
```suggestion
  // Validate SemanticsObject deallocation does not crash.
  // https://github.com/flutter/flutter/issues/66032
```

**Pull request merged**